# Проект урока 3. Ещё больше о Pandas.
#  «Аналитика бронирования отелей»

## Шаг 1

В этом проекте вам предстоит почувствовать себя в роли аналитика данных в компании, владеющей сетью отелей, и проанализировать данные о бронированиях. Описание колонок:
* Hotel – тип отеля (City Hotel или Resort Hotel)  
* Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным
* Lead time – количество дней, прошедших между датой бронирования и датой прибытия  
* Arrival full date – полная дата прибытия
* Arrival date year – год прибытия  
* Arrival date month – месяц прибытия  
* Arrival date week number – номер недели прибытия
* Arrival date day of month – день прибытия
* Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле
* Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле
* Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)
* Adults – число взрослых
* Children – число детей
* Babies – число младенцев
* Meal – выбранный тип питания
* Country – страна происхождения клиента
* Reserved room type – тип зарезервированного номера
* Assigned room type – тип полученного номера (может отличаться от забронированного)
* Customer type – тип бронирования
* Reservation status – значение последнего статуса брони: Canceled — было отменено клиентом; Check-Out — клиент зарегистрировался, но уже покинул отель; No-Show — клиент не зарегистрировался и сообщил администрации отеля причину
* Reservation status date – дата обновления статуса

В этом уроке вы закрепите навыки решения аналитических задач с помощью основных методов библиотеки pandas и попробуете новые методы.

Вы можете решать задания тем способом, который считаете наиболее удобным или оптимальным. А для того, чтобы лучше понять задание, следуйте по шагам выполнения. Бывает, что их можно выполнить в одну строчку, применяя методы друг за другом. А если выполняете шаги отдельно, не забудьте сохранить результат в переменную, чтоьбы в следующем шаге работать именно с ней.

В ноутбуке место для вашего решения обозначено комментарием `# Ваш код здесь`, но вы можете писать код там, где вам удобно, добавлять или удалять ячейки с кодом или текстом по вашему усмотрению.

Напоминаем и о [памятке для новичков в программировании](https://buildin.ai/share/62b112ad-e20b-4b85-ab63-877077d1e9c3?code=2J58AF
)! Очень многие проблемы, описанные в ней, часто встречаются именно в этом уроке, поэтому рекомендуем открыть её в отдельном окне и периодически с ней сверяться.  

Кроме того можно посмотреть подсказки, кликнув на строчку `► Нажмите сюда, чтобы увидеть подсказку`. Также не забывайте о наличии конспектов и возможности задать вопрос в Discord, ссылка на нужный тред есть на странице каждого шага в LMS.

## Шаг 2

Как обычно, в начале откройте датасет и посмотрите на данные. Для этого используйте библиотеку pandas, прочитайте датасет с разделителем `;`, сохраните его в переменную `bookings`. **Посмотрите, из какой страны было сделано бронирование в 19-й сверху строке** (обратите внимание, индекс у этой строки может быть не 19)

Шаги выполнения:

1. Импортируйте библиотеку pandas с общепринятым алиасом (псевдонимом) pd

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>Для импорта библиотеки напишите:

```python
import library as shortcut
```

где вместо `library` подставьте название библиотеки (pandas), а вместо `shortcut` — его общераспространённое сокращённое название (pd)
</p>
</details>

In [1]:
# Ваш код здесь



2. Загрузите csv файл к уроку в свою папку в JupyterНub

3. Прочитайте этот датасет с разделителем `;`, результат сохранив в переменную `bookings`

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Данные находятся в файле формата CSV. Для их загрузки воспользуемся функцией `pd.read_csv()`, передав в качестве аргумента путь к файлу и задав аргумент `sep`, определяющий разделитель. Например,
```python
df = pd.read_csv('data.csv', sep=',')
```
считает файл `data.csv`, хранящийся в той же папке, что и юпитер ноутбук, в котором вы сейчас работаете. Функция учтет, что разделителем в файле является запятая
</p>
</details>

In [2]:
# Ваш код здесь



4. Выведите первые 19 строк датафрейма

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Используйте метод `head()`. Например, код
```python
df.head(10)
```
выведет первые 10 строк датафрейма `df`.
</p>
</details>

In [3]:
# Ваш код здесь



5. Определите название страны из 19-й сверху строки в колонке `Country`

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Обратитесь к названию колонки, взяв его в квадратные скобки и кавычки. Например, код
```python
df.head(10)['col_1']
```
выведет первые 10 строк колонки `col_1` датафрейма `df`.
      
Помните, что вам нужна «19-я сверху строка», это может быть не то же самое, что «строка с индексом 19». Если вы вывели первые 19 строк датафрейма, нужная вам строка будет самой нижней.
</p>
</details>

In [4]:
# Ваш код здесь



**Решение**

In [5]:
import pandas as pd

In [6]:
bookings = pd.read_csv('2_bookings.csv', sep=';')

In [7]:
bookings.head(19)['Country']

0     PRT
1     PRT
2     GBR
3     GBR
4     GBR
5     GBR
6     PRT
7     PRT
8     PRT
9     PRT
10    PRT
11    PRT
12    USA
13    ESP
14    PRT
15    IRL
16    PRT
17    IRL
18    FRA
Name: Country, dtype: object

**Ответ: FRA**

## Шаг 3

Посмотрите на размеры датафрейма `bookings`. **Сколько всего в нем строк и столбцов?**

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Используйте атрибут `shape`, чтобы узнать количество строк и столбцов. Этот атрибут возвращает кортеж, первый элемент которого указывает на количество строк, а второй — на количество столбцов в датафрейме.  

Для обращения к атрибуту датафрейма нужно написать переменную, где хранится датафрейм, поставить точку и написать название атрибута (без скобок!). Например:
```python
df.some_attribute
```
так мы обратимся к атрибуту `some_attribute` у датафрейма в переменной `df`
</p>
</details>

In [8]:
# Ваш код здесь



**Решение**

In [9]:
bookings.shape

(119390, 21)

**Ответ:  
число строк: 119390  
число столбцов: 21**

## Шаг 4

**К какому типу/каким типам относится большинство столбцов?** Выберите один или несколько вариантов.

Шаги выполнения:
1. Проверьте типы данных в датафрейме

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
Есть два способа узнать типы данных:
      
1) Используйте атрибут `dtypes`. Он возвращает серию, индексами которой являются названия столбцов, а значениями — тип данных в этих столбцах.  

Для обращения к атрибуту датафрейма нужно написать переменную, где хранится датафрейм, поставить точку и написать название атрибута (без скобок!). Например:
```python
df.some_attribute
```
Так мы обратимся к атрибуту `some_attribute` у датафрейма в переменной `df`.
      
2) Либо примените к датафрейму метод `info()`, который выводит подробную информацию о датафрейме и его колонках.
      
Как использовать метод? Например, применим метод `some_method()` к датафрейму `df`:
```python
df.some_method()
```
</p>
</details>

In [10]:
# Ваш код здесь



2. Определите, какой тип/какие типы данных преобладают

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Код для этого писать не нужно, посмотрите просто глазами. Если использовали атрибут `dtypes`, количество каждого типа нужно будет посчитать. А метод `info()` все посчитал сам, смотрите на предпоследнюю строчку, которая начинается с `dtypes:`, количество колонок каждого типа указаны в скобках
</p>
</details>

**Решение**

In [11]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [12]:
bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 21 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Hotel                      119390 non-null  object 
 1   Is Canceled                119390 non-null  int64  
 2   Lead Time                  119390 non-null  int64  
 3   arrival full date          119390 non-null  object 
 4   Arrival Date Year          119390 non-null  int64  
 5   Arrival Date Month         119390 non-null  object 
 6   Arrival Date Week Number   119390 non-null  int64  
 7   Arrival Date Day of Month  119390 non-null  int64  
 8   Stays in Weekend nights    119390 non-null  int64  
 9   Stays in week nights       119390 non-null  int64  
 10  stays total nights         119390 non-null  int64  
 11  Adults                     119390 non-null  int64  
 12  Children                   119386 non-null  float64
 13  Babies                     11

**Ответ: int64, object**

## Шаг 5

Как вы уже могли заметить, названия колонок записаны в неудобном для дальнейшей работы формате. **Исправьте ситуацию, приведя названия столбцов к нижнему регистру и заменив пробелы на знак нижнего подчеркивания**.
Например: `Is Canceled` -> `is_canceled`

Вручную это делать долго и неудобно, попробуйте автоматизировать ваше решение. Для этого могут пригодиться:

[rename](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html)  
[str.lower](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.lower.html)  
[str.replace](https://docs.python.org/3/library/stdtypes.html#str.replace)  
[columns](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.columns.html)

Подробнее о них можно прочитать в документации по ссылкам или в конспектах этого и предыдущего уроков.

Шаги выполнения:
1. На одном названии колонки потренируйтесь приводить его к нижнему регистру и заменять пробелы на знак нижнего подчеркивания. Для примера можете взять название колонки `Is Canceled` как строку, в кавычках.

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Чтобы заменить пробелы на нижние подчеркивания, используйте метод `replace()`**, применив его к строке с названием колонки. Этот метод принимает на вход два аргумента, подстроки в таком порядке: что заменить и на что заменить.

**Как использовать метод `replace()`?** Например, применим его к строке `old_str_name`, которая содержит подстроку `old_str`, чтобы заменить её на подстроку `new_str`:
```python
'old_str_name'.replace('old_str', 'new_str')
```

**Чтобы перевести строчку в нижний регистр, воспользуйтесь методом `lower`**. Он не принимает аргументов, а просто переводит всё в нижний регистр, к примеру:
```python
'Hello, World!'.lower()
```
Вернёт строчку, где всё написано маленькими буквами.
      
**Как сразу заменить пробелы и привести буквы к нижнему регистру?** Используйте цепочку методов: напишите два метода подряд, поставив перед каждым точку. Например:
```python
'old_str_name'.some_method_1().some_method_2()
```   
</p>
</details>

In [13]:
# Ваш код здесь



2. Теперь используйте получившийся код для всех названий колонок. Пройдя циклом по списку колонок датафрейма, соберите словарь, в котором ключами будут старые названия колонок, а значениями — новые. Перед циклом, но желательно в той же ячейке, не забудьте создать пустой словарь, куда будете складывать пары названий колонок ключ-значение.

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
    
**Как создать пустой словарь?** Сохраните в переменную пустую пару фигурных скобок. Например:
```python
dictionary_name = {}
```

**Как пройти циклом по списку колонок?** Например:
```python
for col in column_names:
      some action at col
```     
выполнит некоторое действие с каждой колонкой (`col`) в перечне колонок `column_names`.
      
**Для обращения к названиям колонок используйте атрибут `columns`**. Для обращения к атрибуту датафрейма нужно написать переменную, где хранится датафрейм, поставить точку и написать название атрибута (без скобок!). Например:
```python
df.some_attribute
```
так мы обратимся к атрибуту `some_attribute` у датафрейма в переменной `df`.
      
**Как добавить в словарь пару старого и нового названия колонок?** Для этого нужно сохранить новое название, привязав его к ключу - старому названию. Например:

```python
dictionary_name[old_name] = new_name
```
сохранит пару ключ-значение в словарь `dictionary_name`, где ключом является переменная `old_name`, в которой лежит строка со старым названием, а значением - переменная `new_name`, в которой строка с новым названием
</p>
</details>

In [14]:
# Ваш код здесь



3. Убедитесь, что в словаре с названиями колонок получились именно те значения, которые вам нужны

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
    
Как посмотреть содержимое словаря? Запустите ячейку с названием словаря. Например:
```python
dictionary_name
```
выведет содержимое словаря, где запятыми будут разделены пары ключ-значение. Внутри каждой пары сначала идет ключ (старое название), за ним двоеточие, а потом значение (новое название).
</p>
</details>

In [15]:
# Ваш код здесь



4. Используя получившийся словарь, переименуйте колонки датафрейма `bookings`, сохранив результат в том же датафрейме

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Для переименования колонок, передайте в метод `rename()` в параметр `columns` словарь, где ключами являются старые названия колонок, а значениями - новые названия. Например:

```python
df = df.rename(columns=dictionary_name)
```
</p>
</details>

In [16]:
# Ваш код здесь



5. Убедитесь, что названия столбцов были успешно изменены и результат сохранен в датафрейм `bookings`

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Используйте атрибут `columns`. Он возвращает названия колонок датафрейма.  

Для обращения к атрибуту датафрейма нужно написать переменную, где хранится датафрейм, поставить точку и написать название атрибута (без скобок!). Например:
```python
df.some_attribute
```
так мы обратимся к атрибуту `some_attribute` у датафрейма в переменной `df`.
</p>
</details>

In [17]:
# Ваш код здесь



**В LMS датафрейм с прежними названиями колонок сохранен в переменную `bookings`. Перенесите в LMS код с переименованием столбцов и сохранением результата в тот же датафрейм `bookings`. Код из 1, 3 и 5 шагов переносить в LMS не нужно**

**Решение**

In [18]:
'Is Canceled'.replace(' ', '_').lower()

'is_canceled'

In [19]:
rename_cols = {}
for col in bookings.columns:
    rename_cols[col] = col.replace(' ', '_').lower()

In [20]:
rename_cols

{'Hotel': 'hotel',
 'Is Canceled': 'is_canceled',
 'Lead Time': 'lead_time',
 'arrival full date': 'arrival_full_date',
 'Arrival Date Year': 'arrival_date_year',
 'Arrival Date Month': 'arrival_date_month',
 'Arrival Date Week Number': 'arrival_date_week_number',
 'Arrival Date Day of Month': 'arrival_date_day_of_month',
 'Stays in Weekend nights': 'stays_in_weekend_nights',
 'Stays in week nights': 'stays_in_week_nights',
 'stays total nights': 'stays_total_nights',
 'Adults': 'adults',
 'Children': 'children',
 'Babies': 'babies',
 'Meal': 'meal',
 'Country': 'country',
 'Reserved Room Type': 'reserved_room_type',
 'Assigned room type': 'assigned_room_type',
 'customer type': 'customer_type',
 'Reservation Status': 'reservation_status',
 'Reservation status_date': 'reservation_status_date'}

In [21]:
bookings = bookings.rename(columns=rename_cols)

In [22]:
bookings.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_full_date',
       'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'stays_total_nights', 'adults', 'children',
       'babies', 'meal', 'country', 'reserved_room_type', 'assigned_room_type',
       'customer_type', 'reservation_status', 'reservation_status_date'],
      dtype='object')

Код для отправки в LMS:

`rename_cols = {}`  
`for col in bookings.columns:
    rename_cols[col] = col.replace(' ', '_').lower()`  
`bookings = bookings.rename(columns=rename_cols)`

## Шаг 6

Перейдем к исследованию данных. Чтобы вести успешный бизнес, нашей сети отелей нужно понимать своих клиентов. В этом задании выясните, пользователи из каких стран совершили наибольшее число успешных бронирований? Бронирование считается успешным, если в дальнейшем не было отменено (см.колонку `is_canceled`). **В качестве ответа выберите страны, входящие в топ-5 по числу успешных бронирований**.

Шаги выполнения:  
1. Выберите только те записи, где бронирование не было отменено (значение в колонке `is_canceled` равно 0)

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Чтобы отобрать только нужные записи, примените к датафрейму метод `query()`, внутри которого напишите строку с условием, что значение в колонке `is_canceled` равно 0.

Как использовать метод `query()`? Например, отберем строки датафрейма `df`, где в колонке `col_name` значение равно 5:

```python
df.query('col_name == 5')
```   

Учтите, что условие нужно писать в кавычках, а название колонки — без кавычек.
      
Помните, что один знак равно означает присвоение, а два — проверку на равенство.
      
</p>
</details>

In [23]:
# Ваш код здесь



2. Посчитайте количество оставшихся после фильтрации бронирований для каждой страны

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Используйте метод `value_counts()`, применив его к столбцу `country`. Этот метод возвращает серию, где индексы — это уникальные значения из столбца, а значения — это частота их встречаемости. По умолчанию метод располагает записи по убыванию частоты (то есть от большего значения к меньшему)

Как использовать метод? Например применим метод `some_method()` к колонке `column_name` датафрейма `df`:
```python
df.column_name.some_method()
```

</p>
</details>

In [24]:
# Ваш код здесь



3. На основе получившихся расчетов определите 5 стран с наибольшим числом успешных бронирований

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Используйте метод `head()`, который выводит выводит верхние строки датафрейма (по умолчанию — первые 5 строк). Например, код
```python
df.head(10)
```
выведет первые 10 строк датафрейма `df`.
</p>
</details>

In [25]:
# Ваш код здесь



**Решение**

In [26]:
bookings.query('is_canceled==0').country.value_counts().head()

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

**Ответ: DEU, ESP, FRA, GBR, PRT**

## Шаг 7

В нашей сети есть два типа отелей: City Hotel и Resort Hotel. Владельцы сети хотят понимать, отличается ли длительность бронирования в зависимости от типа отеля (см.колонку `hotel`). **Выясните, на сколько ночей (см.колонку `stays_total_nights`) в среднем бронируют отели City Hotel и на сколько Resort Hotel?** Запишите полученные значения в пропуски с точностью до двух знаков после точки.

Может пригодиться: [round](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.round.html)

Шаги выполнения:  
1. Сгруппируйте данные по типу отеля и посчитайте для каждой группы среднее число забронированных ночей

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Сгруппируйте данные по столбцу `hotel`, используя метод `groupby()`**. Его вызов группирует строки по значениям в какой-то колонке (или комбинации значений из нескольких колонок). Например,

```python
df.groupby('month')
```
сгруппирует данные датафрейма `df` по колонке `month`. Сама по себе группировка не вычисляет что-то по группам.

**После группировки используйте метод `mean()` для столбца `stays_total_nights`**. Этот метод посчитает среднее всех значений в указанном столбце.

**Как использовать метод?** Давайте применим метод `some_method()` к колонке `column_name` датафрейма `df`:
```python
df.column_name.some_method()
```

**Как использовать метод сразу после группировки?** Применим метод `some_method()` к колонке `column_name_2` после группировки по столбцу `column_name_1` датафрейма `df`:
```python
df.groupby('column_name_1').column_name_2.some_method()
```      
</p>
</details>

In [27]:
# Ваш код здесь



2. Округлите результаты до двух знаков после точки

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Используйте метод `round()`, который округляет значения во всех или только в указанных колонках датафрейма. По умолчанию округление идет до целого числа. Чтобы задать нужное число знаков после точки, передайте в метод аргумент — целое число. Например применим метод к колонке `column_name` датафрейма `df`, задав округление до пяти знаков после точки:
```python
df.column_name.round(5)
```

</p>
</details>

In [28]:
# Ваш код здесь



**Решение**

In [29]:
bookings.groupby('hotel').stays_total_nights.mean().round(2)

hotel
City Hotel      2.98
Resort Hotel    4.32
Name: stays_total_nights, dtype: float64

**Ответ:  
Среднее число забронированных ночей в отелях типа City Hotel: 2.98  
В Resort Hotel: 4.32**

## Шаг 8

Бывает, что в наших отелях случается [овербукинг](https://ru.wikipedia.org/wiki/Овербукинг). Тогда тип номера, присвоенного клиенту (см.колонку `assigned_room_type`), будет отличаться от изначально забронированного (см.колонку `reserved_room_type`). **Посчитайте, сколько подобных наблюдений встретилось в датасете?** Отмена бронирования также считается.

Шаги выполнения:
1. Отберите строки, где значения в колонке `assigned_room_type` не совпадают со значениями в колонке `reserved_room_type`

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Чтобы отобрать только нужные записи, примените к датафрейму метод `query()`, внутри которого напишите строку с условием, что значение в колонке `assigned_room_type` не равно значению в колонке `reserved_room_type`.

Как использовать метод `query()`? Например, отберем строки датафрейма `df`, где в колонке `col_name` значение равно 5:

```python
df.query('col_name == 5')
```   

Учтите, что условие нужно писать в кавычках, а названия колонок — без кавычек.
      
Помните, что один знак равно означает присвоение, два знака равно — проверку на равенство, а восклицательный знак вместе со знаком равно — проверку на неравенство.
      
</p>
</details>

In [30]:
# Ваш код здесь



2. Посчитайте количество оставшихся строк

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Используйте атрибут `shape`. Он возвращает кортеж, первый элемент которого указывает на количество строк, а второй — на количество столбцов в датафрейме.  

Для обращения к атрибуту датафрейма нужно написать переменную, где хранится датафрейм, поставить точку и написать название атрибута (без скобок!). Например:
```python
df.some_attribute
```
Так мы обратимся к атрибуту `some_attribute` у датафрейма в переменной `df`.
</p>
</details>

In [31]:
# Ваш код здесь



**Решение**

In [32]:
bookings.query('assigned_room_type != reserved_room_type').shape

(14917, 21)

**Ответ: 14917**

## Шаг 9

Если у спроса на услугу есть сезонность, компании стоит это учесть. Проанализируйте даты запланированного прибытия (см.колонку `arrival_date_year`) и **определите, на какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?**

Шаги выполнения:
1. Сгруппируйте данные по годам и посчитайте, сколько раз внутри года встречается каждый месяц

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Сгруппируйте данные по столбцу `arrival_date_year`, используя метод `groupby()`**. Его вызов группирует строки по значениям в какой-то колонке (или комбинации значений из нескольких колонок). Например,
```python
df.groupby('month')
```
сгруппирует данные датафрейма `df` по колонке `month`. Сама по себе группировка не вычисляет что-то по группам.

**После группировки используйте метод `value_counts()`, применив его к столбцу `arrival_date_month`**. Этот метод возвращает серию, где индексы — это уникальные значения из столбца, а значения — это частота их встречаемости. По умолчанию метод располагает записи по убыванию частоты (то есть от большего значения к меньшему).

**Как использовать метод?** Применим метод `some_method()` к колонке `column_name` датафрейма `df`:
```python
df.column_name.some_method()
```

**Как использовать метод сразу после группировки?** Применим метод `some_method()` к колонке `column_name_2` после группировки по столбцу `column_name_1` датафрейма `df`:
```python
df.groupby('column_name_1').column_name_2.some_method()
```      
</p>
</details>

In [33]:
# Ваш код здесь



2. Посмотрите, в каком месяце было больше всего бронирований в 2016 году. А в 2017?

**Решение**

In [34]:
bookings.groupby('arrival_date_year').arrival_date_month.value_counts()

arrival_date_year  arrival_date_month
2015               September             5114
                   October               4957
                   August                3889
                   December              2920
                   July                  2776
                   November              2340
2016               October               6203
                   May                   5478
                   April                 5428
                   September             5394
                   June                  5292
                   August                5063
                   March                 4824
                   July                  4572
                   November              4454
                   February              3891
                   December              3860
                   January               2248
2017               May                   6313
                   April                 5661
                   June                  5

**Ответ:  
В 2016 чаще всего бронировали отель на октябрь.  
В 2017 чаще всего оформляли бронирование на май**.

## Шаг 10

Продолжим изучать бронирования. **Выясните, в каком месяце каждого года бронирования отеля типа City Hotel отменялись чаще всего.**

Шаги выполнения:
1. Оставьте только записи об отмененных бронированиях (см.колонку `is_canceled`) отеля типа City Hotel (см.колонку `hotel`)

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Чтобы отобрать только нужные записи, примените к датафрейму метод `query()`, внутри которого напишите строку с двойным условием: значение в колонке `is_canceled` равно 1 и тип отеля в колонке `hotel` — City Hotel.

Как использовать метод `query()`? Например, отберем строки датафрейма `df`, где в колонке `col_1` значение равно 5 или в колонке `col_2` значение не равно `some_string`:

```python
df.query('col_1 == 5 or col_2 != "some_string"')
```    
      
Между условиями поставьте логический оператор: `and` (или `&`), если оба условия обязательно должны выполняться. Или оператор `or` (или `|`), если должно выполняться хотя бы одно из условий.
      
Учтите, что оба условия нужно писать в единых кавычках, а каждое условие и названия колонок — без кавычек. Целые числа также пишутся без кавычек, а строки — в кавычках (двойных или одинарных — главное, не тех, в которые взяты оба условия).
      
Помните, что один знак равно означает присвоение, а два — проверку на равенство.
      
</p>
</details>

In [35]:
# Ваш код здесь



2. Сгруппируйте оставшиеся данные по годам и посчитайте количество записей для каждого месяца внутри года

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Сгруппируйте данные по столбцу `arrival_date_year`, используя метод `groupby()`**. Его вызов группирует строки по значениям в какой-то колонке (или комбинации значений из нескольких колонок). Например,
```python
df.groupby('month')
```
сгруппирует данные датафрейма `df` по колонке `month`. Сама по себе группировка не вычисляет что-то по группам.

**После группировки используйте метод `value_counts()`, применив его к столбцу `arrival_date_month`**. Этот метод возвращает серию, где индексы — это уникальные значения из столбца, а значения — это частота их встречаемости. По умолчанию метод располагает записи по убыванию частоты (то есть от большего значения к меньшему).

**Как использовать метод?** Применим метод `some_method()` к колонке `column_name` датафрейма `df`:
```python
df.column_name.some_method()
```

**Как использовать метод сразу после группировки?** Применим метод `some_method()` к колонке `column_name_2` после группировки по столбцу `column_name_1` датафрейма `df`:
```python
df.groupby('column_name_1').column_name_2.some_method()
```      
</p>
</details>

In [36]:
# Ваш код здесь



3. Посмотрите, в каком месяце бронирования City Hotel отменялись чаще всего в 2015? 2016? 2017?

**Решение**

In [37]:
bookings.query('hotel == "City Hotel" and is_canceled == 1') \
    .groupby('arrival_date_year') \
    .arrival_date_month \
    .value_counts()

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

**Ответ:  
2015 - September  
2016 - October  
2017 - May**

## Шаг 11

Теперь посмотрим на наших клиентов с точки зрения их возраста. **Какая из колонок `adults`, `children` и `babies` имеет наибольшее среднее значение?**

Шаги выполнения:
1. Оставьте в датафрейме только 3 колонки:  `adults`, `children` и `babies`

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Обратитесь только к трем колонкам датафрейма, передав их список в квадратные скобки. Например,
```python
df[['col1', 'col2']]
```
вернет только две колонки датафрейма `df`. Обратите внимание на двойные квадратные скобки.
   
</p>
</details>

In [38]:
# Ваш код здесь



2. Посчитайте средние значения для оставшихся колонок

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Примените к датафрейму из 3 колонок метод `mean()`, который считает среднее значение.

**Как использовать метод?** Например применим метод `some_method()` к датафрейму `df`:
```python
df.some_method()
```
</p>
</details>

In [39]:
# Ваш код здесь



3. Посмотрите, среднее значение в какой колонке больше других?

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Это можно сделать, посмотрев на значения глазами — код здесь не нужен.
      
Если вас смущает, что получились дробные числа и не может быть условно «полтора человека», то помните, что это среднее значение, поэтому числа могут не быть целыми. В дальнейшем для принятия некоторых бизнес-решений их можно будет округлить до целого, но это требуется не всегда.
</p>
</details>

**Решение**

In [40]:
bookings[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

**Ответ: adults**

## Шаг 12

В наших данных информация о количестве детей записана в две колонки: `children` и `babies`. Это удобно, если нам нужно знать примерный возраст маленьких гостей, но неудобно, если нам требуется просто общее число детей. Для этого создайте колонку `total_kids`, объединив столбцы `children` и `babies`.

После того проверьте, для отелей какого типа среднее значение этой переменной оказалось наибольшим, для City hotel или Resort hotel. **В качестве ответа укажите наибольшее среднее `total_kids`, округлив до двух знаков после точки**.

Шаги выполнения:
1. Создайте колонку `total_kids`, сложив значения в столбцах `children` и `babies`

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Как создать новую колонку?** достаточно написать такой код:
```python
df['new_column'] = some_value
```
где `df` - датафрейм, `new_column` - название новой колонки, `some_value` - то, чему будет равна колонка (это может быть одно число или колонка датафрейма, где число элементов такое же как число строк в `df`)
      
**Как сложить значения в двух колонках?** Арифметические операции с пандасовскимии колонками делаются очень просто - прямо как с питоновскими числами. Например, этот код:
```python
df.column1 * df.column2
```
умножит все значения из колонки `column1` датафрейма `df` на соответствующие элементы колонки `column2`. Число элементов в обеих колонках должно быть одинаковым
</p>
</details>

In [41]:
# Ваш код здесь



2. Сгруппируйте данные по типу отеля и посчитайте для каждой группы среднее количество детей

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Сгруппируйте данные по столбцу `hotel`, используя метод `groupby()`**. Его вызов группирует строки по значениям в какой-то колонке (или комбинации значений из нескольких колонок). Например,

```python
df.groupby('month')
```
сгруппирует данные датафрейма `df` по колонке `month`. Сама по себе группировка не вычисляет что-то по группам.

**После группировки используйте метод `agg()`, в котором примените агрегирующую функцию `mean` к столбцу `total_kids`**. Тогда этот метод посчитает среднее всех значений в указанном столбце.

**Как использовать метод `agg()`?** Например применим его к датафрейму `df` и посчитаем сумму по столбцу `column_name`:
```python
df.agg({'column_name':'sum'})
```

**Как использовать метод `agg()` после группировки?** Используйте цепочку методов, применив их друг за другом. Например:
```python
df.some_method_1().some_method_2()
```      
</p>
</details>

In [42]:
# Ваш код здесь



3. Округлите результаты до двух знаков после точки

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Используйте метод `round()`, который округляет значения во всех или только в указанных колонках датафрейма. По умолчанию округление идет до целого числа. Чтобы задать нужное число знаков после точки, передайте в метод аргумент — целое число. Применим метод к колонке `column_name` датафрейма `df`, задав округление до пяти знаков после точки:
```python
df.column_name.round(5)
```

</p>
</details>

In [43]:
# Ваш код здесь



4. Посмотрите, для отелей какого типа среднее значение оказалось наибольшим? Чему равно это значение?

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Это можно сделать, посмотрев на значения глазами — код здесь не нужен.
      
Если вас смущает, что получились дробные числа и не может быть условно «полтора человека», то помните, что это среднее значение, поэтому числа могут не быть целыми. В дальнейшем для принятия некоторых бизнес-решений их можно будет округлить до целого, но это требуется не всегда.
</p>
</details>

**Решение**

In [44]:
bookings['total_kids'] =  bookings.children + bookings.babies

In [45]:
bookings.groupby('hotel').agg({'total_kids':'mean'}).round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

**Ответ: 0.14**

## Шаг 13

Бизнес не любит упускать прибыль, но не все бронирования завершились успешно (см.колонку `is_canceled`), поэтому попробуем посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитаем метрику под названием Churn Rate.

**Churn rate** (отток, коэффициент оттока) – это процент подписчиков, которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Т.е. представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах. Если Churn rate считают для какой-то определенной группы, то и количество ушедших пользователей, и общее количество пользователей считают только по этой группе.

В нашем случае Churn Rate — это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов. **Посчитайте Churn Rate дважды: для клиентов с детьми и для клиентов без детей. В качестве ответа укажите наибольший Churn Rate (% оттока), округленный до двух знаков после точки** (то есть доля 0.24563 будет 24.56% и в ответ пойдёт 24.56)

Шаги выполнения:  
1. Создайте колонку `has_kids`, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (см.созданную на прошлом шаге колонку `total_kids`), в противном случае – False

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

**Как создать новую колонку?** достаточно написать такой код:
```python
df['new_column'] = some_value
```
где `df` - датафрейм, `new_column` - название новой колонки, `some_value` - то, чему будет равна колонка (это может быть логическое выражение, которое будет истинно или ложно)
      
**Как получить значение True или False?** Логические действия с колонками выполняются по аналогии с арифметическими, просто вместо арифметических операторов - логические. Например, этот код:
```python
df.column1 != df.column2
```
Выдаст колонку с `True` и `False` напротив каждого значения. `True`, если соответствующее значение в колонке `column1` не равно значению из колонки `column2`. Вместо значений в колонке `column2` можно использовать числа или строки
</p>
</details>

In [46]:
# Ваш код здесь



2. Получите 4 датафрейма, отобрав только нужные записи для каждого:  
1 — клиенты без детей, которые делали бронирование (не важно, отменяли его потом или нет)  
2 — клиенты без детей, которые отменили бронирование  
3, 4 — аналогичные два датафрейма для клиентов с детьми

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Чтобы отобрать только нужные записи, примените к датафрейму метод `query()`.

**Как использовать метод `query()`?** Например, отберем строки датафрейма `df`, где в колонке `col_1` значение равно 5 или в колонке `col_2` значение не равно `some_string`:

```python
df.query('col_1 == 5 or col_2 != "some_string"')
```    

**Как написать двойное условие?** Между условиями поставьте логический оператор: `and` (или `&`), если оба условия обязательно должны выполняться. Или оператор `or` (или `|`), если должно выполняться хотя бы одно из условий.
      
Учтите, что оба условия нужно писать в единых кавычках, а каждое условие и названия колонок — без кавычек. Целые числа и логическое значение `True` или `False` также пишутся без кавычек, а строки — в кавычках (двойных или одинарных — главное, не тех, в которые взяты оба условия).

Помните, что один знак равно означает присвоение, два знака равно — проверку на равенство, а восклицательный знак вместе со знаком равно — проверку на неравенство.
      
**Как задать условия?** Для каждого из четырех датафреймов условие будет свое:  
1 — клиенты без детей, которые делали бронирование (значение в колонке `has_kids` будет `False`)  
2 — клиенты без детей, которые отменили бронирование (значение в колонке `is_canceled` равно 1 и значение в колонке `has_kids` будет `False`)  
3 — клиенты с детьми, которые делали бронирование (значение в колонке `has_kids` будет `True`)  
4 — клиенты с детьми, которые отменили бронирование (значение в колонке `is_canceled` равно 1 и значение в колонке `has_kids` будет `True`)
      
</p>
</details>

In [47]:
# Ваш код здесь



3. Посчитайте количество записей в каждом из четырех датафреймов, чтобы получить число клиентов с теми или иными параметрами

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Используйте атрибут `shape`, который возвращает кортеж, первый элемент которого указывает на количество строк, а второй — на количество столбцов в датафрейме.  

Для обращения к элементу атрибута датафрейма нужно написать переменную, где хранится датафрейм, поставить точку и написать название атрибута (без скобок!), после написать номер его элемента в квадратных скобках. Например:
```python
df.some_attribute[1]
```
так мы обратимся ко второму элементу атрибута `some_attribute` у датафрейма `df` (помните, что нумерация элементов начинается с нуля).
</p>
</details>

In [48]:
# Ваш код здесь



4. Посчитайте Churn Rate клиентов без детей: разделите число клиентов без детей, которые отменили бронирование, на общее число клиентов без детей и переведите результат в проценты. По аналогии посчитайте Churn Rate клиентов с детьми

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>
      
Что на что нужно разделить и умножить? Для расчета Churn Rate клиентов без детей разделите число строк 2 датафрейма на число строк 1 датафрейма и умножьте результат на 100 — так получится % оттока.
Для расчета Churn Rate клиентов с детьми разделите число строк 4 датафрейма на число строк 3 датафрейма и умножьте результат на 100.
</p>
</details>

In [49]:
# Ваш код здесь



5. Округлите оба значения Churn Rate до двух знаков после точки

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Используйте `round()` — встроенную функцию python, которая округляет значение поданного в неё числа. По умолчанию округление идет до целого числа. Чтобы задать нужное число знаков после точки, передайте в функцию на втором месте аргумент — целое число. Например применим функцию к числу 12.345, задав округление до 1 знака после точки:
```python
round(12.345, 1)
```
в результате получим число 12.3
</p>
</details>

In [50]:
# Ваш код здесь



6. Посмотрите, какое из значений Churn rate больше: у клиентов с детьми или у клиентов без детей? Чему равно это значение?

<details>
<summary>&#9658; Нажмите сюда, чтобы увидеть подсказку</summary>
  <p>

Это можно сделать, посмотрев на значения глазами — код здесь не нужен.
</p>
</details>

**Решение**

In [51]:
bookings['has_kids'] = bookings['total_kids'] >= 1

In [52]:
# Churn Rate клиентов без детей
round(bookings.query('has_kids==False and is_canceled==1').shape[0] / bookings.query('has_kids==False').shape[0] * 100, 2)

37.22

In [53]:
# Churn Rate клиентов с детьми
round(bookings.query('has_kids==True and is_canceled==1').shape[0] / bookings.query('has_kids==True').shape[0] * 100, 2)

34.92

**Ответ: 37.22**

## Итоги

В этом проекте вы проанализировали данные о бронированиях и закрепили навыки работы с библиотекой pandas, а также узнали новые подходы к переименованию колонок и округлению значений, познакомились с метрикой Churn Rate.